STULabel/STUMainScreenProperties.m

```diff
@@ -57,31 +57,32 @@ static void updateMainScreenProperties(void) {
 #undef store
 }
 
-@interface UIScreen (STUMainScreenProperties)
-+ (void)load;
-@end
-@implementation UIScreen (STUMainScreenProperties)
-+ (void)load {
-  // We can't do this initialization lazily, because UIScreen must only be accessed on the
-  // main thread. (Using `dispatch_sync(dispatch_get_main_queue(), ...)` would lead to a
-  // deadlock when the main thread is waiting for the thread in which stu_mainScreen... is called
-  // for the first time.)
```

Comment:
Does the proposed change avoid this pitfall?

Reply:
Ah, perhaps not. I originally took the comment to be warning against dispatch_sync targeting the current queue. But now I figure you're also worried about the main thread e.g. waiting on some lock that only gets released by some other thread after it checks main screen properties?

Other ideas that occur to me:

• UIApplicationDidFinishLaunchingNotification. Too late.
• dispatch_async(dispatch_get_main_queue()) from +load. I don't know when that will run exactly but it's after main(). I'm also not sure whether it's safe to assume the main queue is available in +load.
• Check whether we're running in a unit test bundle and delay initialization until… sometime? XCTest emits some notifications and has some delegate methods but no obvious candidates jumped out at me.
• Add some kind of +[STULabel loadMainScreenProperties] method and ask library users to call it early from the main thread.
• Switch to lazy-loading and ask STULabel users not to park the main thread and make calls elsewhere when main screen properties may not have been initialized.
• Dangle off some UIKit code that runs early but not too early (e.g. in +load we might swizzle something on UIApplication).

I'm not sure what set of tradeoffs you're looking for. I believe it's currently possible to call stu_mainScreen… functions before initialization from another category +load method (assuming the order they're called within the same binary is undefined) so it just depends on where you wanna draw the line.

Reply:
> Add some kind of +[STULabel loadMainScreenProperties] method and ask library users to call it early from the main thread.

This is the approach we are taking for now in our fork. It’s not ideal, but neither are any of the alternatives (I can’t think of any others than that hat you mentioned)

I don’t know the implementation well enough to understand why these properties are needed off of the main thread. Perhaps the authors could chime in. Seems like if we could rely on main thread access this could be fixed. Maybe something that’s currently synchronous needs to be async.

Reply:
>> // We can't do this initialization lazily, because UIScreen must only be accessed on the
>> // main thread. (Using `dispatch_sync(dispatch_get_main_queue(), ...)` would lead to a
>> // deadlock when the main thread is waiting for the thread in which stu_mainScreen... is called
>> // for the first time.)

> I originally took the comment to be warning against dispatch_sync targeting the current queue. But now I figure you're also worried about the main thread e.g. waiting on some lock that only gets released by some other thread after it checks main screen properties?

`dispatch_sync(dispatch_get_main_queue(), ...)` will not just deadlock when executed on the main thread itself, but also when the main thread is waiting for the current thread (on which `dispatch_sync` is being called).

If Apple doesn't make `UIScreen` properties safe to call at library load time, then the only solution I see is adding an explicit library initialization method for STULabel that must be called from the main thread.